### PR TITLE
okx: add stop order in editOrder

### PR DESCRIPTION
**Pull request by sc0Vu**
okx: add stop order in editOrder

### CODE REVIEW

File: ts/src/okx.ts [hunk 1]
```diff
@@ -2298,33 +2298,73 @@ export default class okx extends Exchange {
          * @param {float} amount how much of the currency you want to trade in units of the base currency
          * @param {float|undefined} price the price at which the order is to be fullfilled, in units of the quote currency, ignored in market orders
          * @param {object} params extra parameters specific to the okx api endpoint
+         * @param {bool} params.stop True if is algo order
```
Comment (carlosmiei) [inline ts/src/okx.ts:2301]: We should add here all the other custom parameters such as `takeProfitPrice`, `stopLossPrice`, etc

File: ts/src/okx.ts [hunk 2]
```diff
@@ -2298,33 +2298,73 @@ export default class okx extends Exchange {
...
-        params = this.omit (params, [ 'clOrdId', 'clientOrderId' ]);
+        params = this.omit (params, [ 'clOrdId', 'clientOrderId', 'algoClOrdId' ]);
         if (amount !== undefined) {
             request['newSz'] = this.amountToPrecision (symbol, amount);
         }
-        if (price !== undefined) {
-            request['newPx'] = this.priceToPrecision (symbol, price);
-        }
-        const response = await (this as any).privatePostTradeAmendOrder (this.extend (request, params));
+        const response = await (this as any)[method] (this.extend (request, params));
```
Comment (carlosmiei) [inline ts/src/okx.ts:2367]: (this is any) is not needed

File: ts/src/okx.ts [hunk 3]
```diff
@@ -2298,33 +2298,73 @@ export default class okx extends Exchange {
...
-        if (clientOrderId !== undefined) {
-            request['clOrdId'] = clientOrderId;
+        const isStop = this.safeValue (params, 'stop');
+        const tpTriggerPx = this.safeValue2 (params, 'takeProfitPrice', 'tpTriggerPx');
+        const tpOrdPx = this.safeValue (params, 'tpOrdPx', price);
+        const tpTriggerPxType = this.safeString (params, 'tpTriggerPxType');
+        const slTriggerPx = this.safeValue2 (params, 'stopLossPrice', 'slTriggerPx');
+        const slOrdPx = this.safeValue (params, 'slOrdPx', price);
+        const slTriggerPxType = this.safeString (params, 'slTriggerPxType');
+        const conditional = (tpTriggerPx !== undefined) || (slTriggerPx !== undefined) || (type === 'conditional');
```
Comment (carlosmiei) [inline ts/src/okx.ts:2319]: shouldn't we check `tpOrderPx`, `slOrderPx` as well here?

Comment (sc0Vu) [reply ts/src/okx.ts:2326]: @carlosmiei the default value for `slOrdPx` & `tpOrdPx` is price, maybe we don't need to check these value for conditional?

Comment (carlosmiei) [reply ts/src/okx.ts:2326]: yeah I think so

File: ts/src/okx.ts [hunk 4]
```diff
@@ -2291,40 +2291,87 @@ export default class okx extends Exchange {
...
         };
-        const clientOrderId = this.safeString2 (params, 'clOrdId', 'clientOrderId');
-        if (clientOrderId !== undefined) {
-            request['clOrdId'] = clientOrderId;
+        const isStop = this.safeValue (params, 'stop');
+        const tpTriggerPx = this.safeValue2 (params, 'takeProfitPrice', 'tpTriggerPx');
+        const tpOrdPx = this.safeValue (params, 'tpOrdPx', price);
+        const tpTriggerPxType = this.safeString (params, 'tpTriggerPxType');
+        const slTriggerPx = this.safeValue2 (params, 'stopLossPrice', 'slTriggerPx');
+        const slOrdPx = this.safeValue (params, 'slOrdPx', price);
```
Comment (carlosmiei) [inline ts/src/okx.ts:2324]: @sc0Vu why safeValue here? we should probably use `safeNumber` no?

File: ts/src/okx.ts [hunk 5]
```diff
@@ -2291,40 +2291,87 @@ export default class okx extends Exchange {
...
+            }
+            if (slOrdPx !== undefined) {
+                request['newSlOrdPx'] = this.priceToPrecision (symbol, slOrdPx);
+            }
+            if (slTriggerPxType !== undefined) {
+                request['newSlTriggerPxType'] = slTriggerPxType;
+            }
         } else {
-            request['ordId'] = id;
+            if (!market['spot']) {
```
Comment (carlosmiei) [inline ts/src/okx.ts:2357]: @sc0Vu I just tested and it seems feasible to edit swap orders
image

File: ts/src/okx.ts [hunk 6]
```diff
@@ -2291,40 +2291,84 @@ export default class okx extends Exchange {
...
+        const conditional = (tpTriggerPx !== undefined) || (slTriggerPx !== undefined) || (type === 'conditional');
+        const clientOrderId = this.safeStringN (params, [ 'clOrdId', 'clientOrderId', 'algoClOrdId' ]);
+        let method = undefined;
+        // Support stop order only
+        if (conditional || isStop) {
+            // Only applicable to Futures and Perpetual swap.
+            if ((!market['future']) && !market['swap']) {
+                throw new NotSupported (this.id + ' editOrder() does not support ' + market['type'] + ' orders, only future & swap orders are accepted');
+            }
+            method = 'privatePostTradeAmendAlgos';
```
[the remark this answers is not in the document]
Comment (sc0Vu) [reply ts/src/okx.ts:2335]: @carlosmiei yes, I agree. But I didn't see triggerPx in api params... Will check with them.

File: ts/src/okx.ts [hunk 7]
```diff
@@ -2445,40 +2445,84 @@ export default class okx extends Exchange {
...
+        const slTriggerPx = this.safeString2 (params, 'stopLossPrice', 'slTriggerPx');
+        const slOrdPx = this.safeString (params, 'slOrdPx', price);
+        const slTriggerPxType = this.safeString (params, 'slTriggerPxType');
+        const conditional = (tpTriggerPx !== undefined) || (slTriggerPx !== undefined) || (type === 'conditional');
+        const clientOrderId = this.safeStringN (params, [ 'clOrdId', 'clientOrderId', 'algoClOrdId' ]);
+        let method = undefined;
+        // Support stop order only
+        if (conditional || isStop) {
+            // Only applicable to Futures and Perpetual swap.
+            if ((!market['future']) && !market['swap']) {
```
Comment (carlosmiei) [inline ts/src/okx.ts:2486]: stopOrders/tk/sl editing not working for some reason
```
CCXT v3.1.15
okx.editOrder (583769858395947008, LTC/USDT:USDT, limit, buy, 2, 55, [object Object])
BadRequest okx {"code":"1","data":[{"algoClOrdId":"","algoId":"583769858395947008","reqId":"","sCode":"51000","sMsg":"Parameter newTpTriggerPxType  error"}],"msg":""}
---------------------------------------------------
[BadRequest] okx {"code":"1","data":[{"algoClOrdId":"","algoId":"583769858395947008","reqId":"","sCode":"51000","sMsg":"Parameter newTpTriggerPxType  error"}],"msg":""}

    at throwExactlyMatchedException  Users/cjg/Git/ccxt10/ccxt/js/src/base/Exchange.js:2851  
    at handleErrors                  Users/cjg/Git/ccxt10/ccxt/js/src/okx.js:6382            
    at                               Users/cjg/Git/ccxt10/ccxt/js/src/base/Exchange.js:748   
    at processTicksAndRejections     node:internal/process/task_queues:95                    
    at fetch2                        Users/cjg/Git/ccxt10/ccxt/js/src/base/Exchange.js:2492  
    at request                       Users/cjg/Git/ccxt10/ccxt/js/src/base/Exchange.js:2495  
    at editOrder                     Users/cjg/Git/ccxt10/ccxt/js/src/okx.js:2516            
    at async run                     Users/cjg/Git/ccxt10/ccxt/examples/js/cli.js:295        

okx {"code":"1","data":[{"algoClOrdId":"","algoId":"583769858395947008","reqId":"","sCode":"51000","sMsg":"Parameter newTpTriggerPxType  error"}],"msg":""}
```

Comment (carlosmiei) [reply ts/src/okx.ts:2486]: @sc0Vu any clue?

**Comment by carlosmiei** [comment]
continues here https://github.com/ccxt/ccxt/pull/20970

**Pull request closed without merging**